pkg/extractor/extractor.go

```diff
@@ -0,0 +1,158 @@
+package extractor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	imageparser "github.com/novln/docker-parser"
+	"github.com/operator-framework/operator-registry/pkg/registry"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+type MainExtractor struct {
+	Log    logrus.FieldLogger
+	Index  *DefaultIndexExtractor
+	Bundle *DefaultBundleExtractor
```

Comment:
Any particular reason to define concrete types here? I thought this is why you made the `IndexExtractor` and `BundleExtractor` interfaces. So different implementations can just be plugged in.

Reply:
I wanted to use the interface types, but I get issues when trying to pass the logger down in the structs. Any ideas?

Reply:
The cache and encoder components of both index/bundle extractor are still configurable, so it's still pretty easy to plug something different. Example:
```go
encoder := NewProtobufSnappyBundleEncoder()
cache := NewRedisBundleCache(encoder)
bundleExtractor := NewBundleExtractor(WithCache(cache))
extractor := extractor.New(WithBundleExtractor(bundleExtractor))
```

Reply:
This is how I would inject dependencies: https://github.com/Ajpantuso/addon-metadata-operator/commit/0b51ffafcc8329d8afa41f60ecad43a34fc48bd8

Reply:
refactored